File: src/tfbase/Sounds.py
"""
Builds the sound list from the script file.
"""

import math
from enum import IntEnum
from panda3d.core import (
    ConfigVariableDouble, ConfigVariableList, Filename, KeyValues,
    PStatCollector, SteamAudioProperties, loadPrcFileData, ProxyAudioSound)
import random

from tf.tfbase import TFGlobals

csc_coll = PStatCollector("App:Sounds:CreateSoundClient")
get_sound_coll = PStatCollector("App:Sounds:CreateSoundClient:GetSound")

snd_refdb = ConfigVariableDouble("snd-refdb", 60.0)
snd_refdist = ConfigVariableDouble("snd-refdist", 36)

load_sounds = ConfigVariableList("load-sounds")

def attenToSoundLevel(attn):
    return ((50 + 20 / float(attn)) if attn else 0.0)

def soundLevelToAtten(sndlvl):
    return ((20.0 / float(sndlvl - 50)) if sndlvl > 50 else 4.0)

def attenToDistMult(attn):
    return soundLevelToDistMult(attenToSoundLevel(attn))

def soundLevelToDistMult(sndlvl):
    if sndlvl:
        return (math.pow(10.0, snd_refdb.value / 20) / math.pow(10.0, float(sndlvl) / 20)) / snd_refdist.value
    else:
        return 0.0

def distMultToSoundLevel(distMult):
    if distMult:
        return 20 * math.log10(math.pow(10.0, snd_refdb.value / 20) / (distMult * snd_refdist.value))
    else:
        return 0.0

def setSoundLevel(sound, sndlvl):
    sound.set3dMinDistance(1.0 / soundLevelToDistMult(sndlvl))

def sourcePitchToPlayRate(pitch):
    if pitch < 100:
        return TFGlobals.remapVal(pitch, 0, 100, 0.5, 1.0)
    elif pitch > 100:
        return TFGlobals.remapVal(pitch, 100, 255, 1.0, 2.0)
    else:
        return 1.0

class Channel(IntEnum):

    Invalid = -1

    CHAN_AUTO = 0
    CHAN_WEAPON = 1
    CHAN_VOICE = 2
    CHAN_ITEM = 3
    CHAN_BODY = 4
    CHAN_STREAM = 5
    CHAN_STATIC = 6
    CHAN_VOICE_BASE = 7
    CHAN_VOICE2 = 8

class SoundLevel(IntEnum):

    SNDLVL_NONE = 0
    SNDLVL_IDLE = 60
    SNDLVL_TALKING = 80
    SNDLVL_NORM = 75
    SNDLVL_STATIC = 66
    SNDLVL_GUNFIRE = 140

class Pitch(IntEnum):

    PITCH_LOW = 95
    PITCH_NORM = 100
    PITCH_HIGH = 120

class Wave:

    def __init__(self):
        self.filename = ""
        # Cached AudioSound data, None until first time wave is needed.
        self.sound = None
        self.spatialized = False
        self.loopStart = 0.0
        self.loopEnd = -1.0
        self.asyncHandle = None
        self.asyncProxies = []

    def __asyncCallback(self, sound):
        self.sound = sound
        self.sound.setLoopRange(self.loopStart, self.loopEnd)

        # Copy sound to each proxy.
        for proxy, mgr in self.asyncProxies:
            proxy.setRealSound(mgr.getSound(self.sound))
        self.asyncProxies = []

        self.asyncHandle = None

    def getSound(self, mgr):
        if self.asyncHandle:
            assert self.asyncProxies
            proxy = ProxyAudioSound(self.asyncProxies[0][0])
            self.asyncProxies.append((proxy, mgr))
            return proxy

        elif self.sound:
            return mgr.getSound(self.sound)

        else:
            proxy = ProxyAudioSound()
            proxy.setLoopRange(self.loopStart, self.loopEnd)
            self.asyncProxies = [(proxy, mgr)]
            self.asyncHandle = base.loader.loadSound(mgr, self.filename.getFullpath(),
                self.spatialized, self.__asyncCallback)
            return proxy

    def __repr__(self):
        return f"""
        Wave {self.filename}
          sound: {self.sound}
          spatialized: {self.spatialized}
        """

class SoundInfo:
    """
    Info about a single sound from the script file.
    """

    def __init__(self):
        # The name of it in the script file, for instance "Weapon_Wrench.Draw"
        self.name = ""
        # Volume range
        self.volume = [1.0, 1.0]
        # Pitch range.
        self.pitch = [1.0, 1.0]
        # The channel on the SoundEmitter that the sound plays on.
        # If not CHAN_STATIC or CHAN_AUTO, overrides the currently playing sound on that
        # channel.
        self.channel = Channel.CHAN_AUTO
        self.soundLevel = 0
        # This changes the spatial attenuation of the sound, comes from SNDLVL_X
        self.minDistance = 1.0
        self.distMult = 1.0
        self.wave = None
        # If there are multiple, we will randomly choose one.
        self.waves = []
        self.index = 0

    def __repr__(self):
        return f"""
        Sound {self.name}
          volume:     [{self.volume[0]}, {self.volume[1]}]
          channel:    {self.channel}
          soundLevel: {self.soundLevel}
          minDist:    {self.minDistance}
          wave:       {self.wave}
          distMult:   {self.distMult}
          waves       {self.waves}
        """

Sounds = {}
AllSounds = []

def createSound(info, spatial=False, getWave=False):
    if not info:
        return None

    wave = None
    if info.wave:
        wave = info.wave
    elif len(info.waves) > 0:
        wave = random.choice(info.waves)
    if not wave:
        return None

    sound = wave.getSound(base.sfxManager)
    sound.setPlayRate(random.uniform(info.pitch[0], info.pitch[1]))
    sound.setVolume(random.uniform(info.volume[0], info.volume[1]))
    if spatial:
        sound.set3dMinDistance(info.minDistance)
        props = SteamAudioProperties()
        props._enable_occlusion = True
        #props._enable_air_absorption = False
        sound.applySteamAudioProperties(props)
        #print(repr(info))

    if getWave:
        return (sound, wave)
    else:
        return sound

def createSoundByName(name, getInfo=False, spatial=False):
    info = Sounds.get(name, None)
    if not getInfo:
        return createSound(info, spatial)
    else:
        return (createSound(info, spatial), info)

def createSoundByIndex(index, getInfo=False, spatial=False):
    info = AllSounds[index]
    if not getInfo:
        return createSound(info, spatial)
    else:
        return (createSound(info, spatial), info)

def createSoundClient(index, waveIndex, volume, pitch, spatialized = False, getInfo=False):
    csc_coll.start()

    if index >= len(AllSounds):
        csc_coll.stop()
        if getInfo:
            return (None, None)
        return None

    info = AllSounds[index]
    if info.wave:
        wave = info.wave
    else:
        wave = info.waves[waveIndex]

    if not wave:
        csc_coll.stop()
        if getInfo:
            return (None, None)
        return None

    get_sound_coll.start()
    sound = wave.getSound(base.sfxManager)
    get_sound_coll.stop()
    sound.setPlayRate(pitch)
    sound.setVolume(volume)
    if spatialized:
        sound.set3dMinDistance(info.minDistance)
        props = SteamAudioProperties()
        props._enable_occlusion = True
        #props._enable_air_absorption = False
        sound.applySteamAudioProperties(props)
        #print(repr(info))

    csc_coll.stop()
    if getInfo:
        return (sound, info)
    return sound

def createSoundServer(name):
    info = Sounds.get(name, None)
    if not info:
        return None

    waveIdx = -1
    if info.wave:
        waveIdx = 0
    else:
        waveIdx = random.randint(0, len(info.waves) - 1)
    if waveIdx == -1:
        return None

    return [info.index, waveIdx, random.uniform(info.volume[0], info.volume[1]),
            random.uniform(info.pitch[0], info.pitch[1]), info.channel]

def processSound(kv):
    global Sounds

    info = SoundInfo()
    info.name = kv.getName()
    for i in range(kv.getNumKeys()):
        key = kv.getKey(i)
        value = kv.getValue(i)

        if key == "channel":
            info.channel = Channel[value]
        elif key == "volume":
            if value == "VOL_NORM":
                info.volume = [1.0, 1.0]
            else:
                minmax = "".join(value.split()).split(",")
                if len(minmax) > 1:
                    info.volume = [float(minmax[0]), float(minmax[1])]
                else:
                    info.volume = [float(minmax[0]), float(minmax[0])]
        elif key == "soundlevel":
            dbs = value.lower().split("sndlvl_")[1]
            if dbs.endswith("db"):
                info.soundLevel = int(dbs[:len(dbs) - 2])
            elif dbs.endswith("dbm"):
                # Fuckers.
                info.soundLevel = int(dbs[:len(dbs) - 3])
            else:
                info.soundLevel = SoundLevel[value]
            info.distMult = soundLevelToDistMult(info.soundLevel)
            if info.distMult > 0.0:
                info.minDistance = 1.0 / info.distMult
            else:
                info.minDistance = 1000000.0
        elif key == "wave":
            info.wave = Wave()
            if value.startswith(")") or value.startswith(">") or value.startswith("<"):
                info.wave.spatialized = True
                value = value[1:]
            elif value.startswith("#"):
                value = value[1:]
            info.wave.filename = Filename.fromOsSpecific(value.lower())
        elif key == "loopstart":
            assert info.wave
            info.wave.loopStart = float(value)
        elif key == "loopend":
            assert info.wave
            info.wave.loopEnd = float(value)
        elif key == "pitch":
            if value.upper() in Pitch.__members__:
                info.pitch = [Pitch[value.upper()], Pitch[value.upper()]]
            else:
                minmax = "".join(value.split()).split(",")
                if len(minmax) > 1:
                    info.pitch = [float(minmax[0]), float(minmax[1])]
                else:
                    info.pitch = [float(minmax[0]), float(minmax[0])]

            info.pitch[0] = sourcePitchToPlayRate(info.pitch[0])
            info.pitch[1] = sourcePitchToPlayRate(info.pitch[1])

    for i in range(kv.getNumChildren()):
        child = kv.getChild(i)
        if child.getName() == "rndwave":
            for j in range(child.getNumKeys()):
                value = child.getValue(j)
                wave = Wave()
                if value.startswith(")") or value.startswith(">") or value.startswith("<"):
                    wave.spatialized = True
                    value = value[1:]
                elif value.startswith("#"):
                    value = value[1:]
                wave.filename = Filename.fromOsSpecific(value.lower())
                info.waves.append(wave)

    info.index = len(AllSounds)
    Sounds[info.name] = info
    AllSounds.append(info)

def loadSounds(server = False):
    for i in range(load_sounds.getNumUniqueValues()):
        filename = Filename.fromOsSpecific(load_sounds.getUniqueValue(i))

        print("Loading sounds from %s" % filename.getFullpath())

        kv = KeyValues.load(filename)
        if not kv:
            print("Unable to load sound script file", filename)
            continue

        for j in range(kv.getNumChildren()):
            child = kv.getChild(j)
            processSound(child)

    #print(repr(Sounds))

def incRefDb():
    db = snd_refdb.value
    db += 1
    loadPrcFileData('', 'snd-refdb %s' % db)
    print("ref db:", db)

def decRefDb():
    db = snd_refdb.value
    db -= 1
    loadPrcFileData('', 'snd-refdb %s' % db)
    print("ref db:", db)

def enableRefDbDebug():
    base.accept('arrow_up', incRefDb)
    base.accept('arrow_down', decRefDb)

File: src/cxxapp/client/sounds.cxx
#include "sounds.h"
#include "audioManager.h"
#include "audioSound.h"
#include "config_audio.h"
#include "virtualFileSystem.h"
#include "configVariableList.h"
#include "keyValues.h"
#include "notifyCategoryProxy.h"
#include "string_utils.h"
#include "audioEngine.h"
#include <algorithm>
#include <cctype>
#include <cmath>
#include <cstddef>
#include "nullAudioSound.h"
#include "steamAudioProperties.h"
#include "audioLoadRequest.h"
#include "loader.h"
#include "proxyAudioSound.h"

NotifyCategoryDeclNoExport(sounds);
NotifyCategoryDef(sounds, "tf");

static ConfigVariableList load_sounds_var
("load-sounds",
 PRC_DESC("Sounds script filenames to load at game start."));
static ConfigVariableDouble snd_refdb
("snd-refdb", 60.0,
 PRC_DESC("Reference db for calculating sound distance attenuation."));
static ConfigVariableDouble snd_refdist
("snd-refdist", 36);

static ConfigVariableDouble master_volume
("master-volume", 1.0f,
 PRC_DESC("Master game volume, both sfx and music."));
static ConfigVariableDouble sfx_volume
("sfx-volume", 1.0f,
 PRC_DESC("Volume of sound effects."));
static ConfigVariableDouble music_volume
("music-volume", 1.0f,
 PRC_DESC("Volume of music."));

SoundManager *SoundManager::_global_ptr = nullptr;

static PT(NullAudioSound) null_sound = new NullAudioSound;

/**
 *
 */
static SoundManager::SoundChannel
channel_from_script_name(const std::string &name) {
  if (name == "CHAN_AUTO") {
    return SoundManager::CHANNEL_auto;
  } else if (name == "CHAN_WEAPON") {
    return SoundManager::CHANNEL_weapon;
  } else if (name == "CHAN_VOICE") {
    return SoundManager::CHANNEL_voice;
  } else if (name == "CHAN_ITEM") {
    return SoundManager::CHANNEL_item;
  } else if (name == "CHAN_BODY") {
    return SoundManager::CHANNEL_body;
  } else if (name == "CHAN_STREAM") {
    return SoundManager::CHANNEL_stream;
  } else if (name == "CHAN_STATIC") {
    return SoundManager::CHANNEL_static;
  } else if (name == "CHAN_VOICE_BASE") {
    return SoundManager::CHANNEL_voice_base;
  } else if (name == "CHAN_VOICE2") {
    return SoundManager::CHANNEL_voice_2;
  } else {
    return SoundManager::CHANNEL_invalid;
  }
}

/**
 *
 */
static int
soundlevel_from_script_name(const std::string &name) {
  // Could be SNDLEVEL_TALKING, SNDLVL_55db, SNDLEVEL_65dbm.
  std::string dbs = downcase(name);
  // Get everything after the sndlvl_
  size_t sndlvl_pos = dbs.find("sndlvl_");
  if (sndlvl_pos == std::string::npos) {
    return SoundManager::SOUNDLEVEL_none;
  }
  dbs = dbs.substr(sndlvl_pos + 7);
  size_t len = dbs.length();
  if (dbs.length() > 2 && dbs.substr(dbs.length() - 2) == "db") {
    // SNDLVL_XXdb
    int ret;
    if (!string_to_int(dbs.substr(0, dbs.length() - 2), ret)) {
      return SoundManager::SOUNDLEVEL_none;
    }
    return ret;

  } else if (dbs.length() > 3 && dbs.substr(dbs.length() - 3) == "dbm") {
    // SNDLVL_XXdbm
    int ret;
    if (!string_to_int(dbs.substr(0, dbs.length() - 3), ret)) {
      return SoundManager::SOUNDLEVEL_none;
    }
    return ret;

  } else {
    // It's one of the named soundlevels.
    if (dbs == "none") {
      return SoundManager::SOUNDLEVEL_none;
    } else if (dbs == "idle") {
      return SoundManager::SOUNDLEVEL_idle;
    } else if (dbs == "talking") {
      return SoundManager::SOUNDLEVEL_talking;
    } else if (dbs == "norm") {
      return SoundManager::SOUNDLEVEL_norm;
    } else if (dbs == "static") {
      return SoundManager::SOUNDLEVEL_static;
    } else if (dbs == "gunfire") {
      return SoundManager::SOUNDLEVEL_gunfire;
    } else {
      return SoundManager::SOUNDLEVEL_none;
    }
  }
}

/**
 *
 */
static int
pitch_from_script_name(const std::string &value) {
  std::string lval = downcase(value);
  if (lval == "pitch_low") {
    return SoundManager::PITCH_low;
  } else if (lval == "pitch_norm") {
    return SoundManager::PITCH_norm;
  } else if (lval == "pitch_high") {
    return SoundManager::PITCH_high;
  } else {
    return SoundManager::PITCH_invalid;
  }
}

/**
 *
 */
static SoundManager::Wave
wave_from_script(const std::string &value) {
  SoundManager::Wave wv;
  std::string fname = downcase(value);
  if (!fname.empty() && (fname[0] == ')' || fname[0] == '>' || fname[0] == '<')) {
    wv.spatialized = true;
    fname = fname.substr(1);
  } else if (!fname.empty() && fname[0] == '#') {
    fname = fname.substr(1);
  }
  wv.filename = Filename::from_os_specific(fname);
  return wv;
}

/**
 *
 */
static float
soundlevel_to_dist_mult(int soundlevel) {
  if (soundlevel > 0) {
    return (powf(10.0f, (float)snd_refdb / 20.0f) / powf(10.0f, (float)soundlevel / 20.0f)) / (float)snd_refdist;
  } else {
    return 0.0f;
  }
}

/**
 *
 */
static int
dist_mult_to_soundlevel(float dist_mult) {
  if (dist_mult > 0.0f) {
    return 20 * log10f(powf(10.0f, (float)snd_refdb / 20.0f) / (dist_mult * (float)snd_refdist));
  } else {
    return SoundManager::SOUNDLEVEL_none;
  }
}

/**
 *
 */
static float
pitch_to_play_rate(float pitch) {
  return pitch * 0.01f;
}

/**
 *
 */
static float
db_to_gain(float db) {
  return powf(10.0f, db / 20.0f);
}

/**
 *
 */
static float
gain_to_db(float gain) {
  return 20.0f * logf(gain);
}

/**
 *
 */
static int
atten_to_soundlevel(float atten) {
  return atten > 0 ? ((50 + 20 / (float)atten)) : 0;
}

/**
 *
 */
static float
soundlevel_to_atten(int soundlevel) {
  return soundlevel > 50 ? (20.0f / (float)(soundlevel - 50)) : 4.0f;
}

/**
 *
 */
static std::string
remove_spaces(const std::string &str) {
  std::string trimmed = str;
  trimmed.erase(std::remove_if(trimmed.begin(), trimmed.end(), ::isspace), trimmed.end());
  return trimmed;
}

/**
 *
 */
void SoundManager::
initialize() {
  _audio_engine = AudioEngine::make_engine();

  // An audio manager is essentially a mixing group in a hierarchy.

  _music_manager = _audio_engine->make_manager("music");
  _music_manager->set_volume(music_volume * master_volume);
  // Allow only one song to play at a time.  We could also just manage this ourselves.
  _music_manager->set_concurrent_sound_limit(1);

  _sfx_manager = _audio_engine->make_manager("sfx");
  _sfx_manager->set_volume(sfx_volume * master_volume);
}

/**
 *
 */
void SoundManager::
update() {
  // Engine update automatically updates all managers.
  _audio_engine->update();

  // Check if any wave async loads finished.
  pvector<Wave *> done_waves;
  for (Wave *wv : _loading_waves) {
    assert(wv->load_req != nullptr);
    if (wv->load_req->is_ready()) {
      if (sounds_cat.is_debug()) {
	sounds_cat.debug()
	  << "Async load done for wave " << wv->filename << ", frame "
	  << ClockObject::get_global_clock()->get_frame_count() << "\n";
      }
      // Cool, the sound is ready.
      // Copy it to each proxy that got created during the load.
      wv->sound = wv->load_req->get_sound();
      wv->sound->set_loop_range(wv->loop_start, wv->loop_end);
      for (AsyncProxy &proxy : wv->async_proxies) {
	proxy.proxy->set_real_sound(proxy.mgr->get_sound(wv->sound));
      }
      wv->async_proxies.clear();
      wv->load_req = nullptr;
      done_waves.push_back(wv);
    }
  }
  // Now clear out the waves that finished from the loading list.
  for (Wave *wv : done_waves) {
    _loading_waves.erase(wv);
  }
}

/**
 * Loads all sound script files specified in the load-sounds config variable.
 */
void SoundManager::
load_sounds() {
  for (size_t i = 0; i < load_sounds_var.get_num_unique_values(); ++i) {
    const std::string &val = load_sounds_var.get_unique_value(i);
    Filename fname = Filename::from_os_specific(val);
    load_sound_script_file(fname);
  }
}

/**
 * Loads a single sound script file.
 */
void SoundManager::
load_sound_script_file(const Filename &filename) {
  PT(KeyValues) kv = KeyValues::load(filename);
  if (kv == nullptr) {
    sounds_cat.warning()
      << "Unable to load sound script file " << filename << "\n";
    return;
  }

  sounds_cat.info()
      << "Loading sounds from " << filename << "\n";

  for (size_t i = 0; i < kv->get_num_children(); ++i) {
    process_sound(kv->get_child(i));
  }
}

/**
 * Prints info about all loaded sounds to the console.
 */
void SoundManager::
list_sounds() const {
  sounds_cat.info()
    << "Num sounds: " << _sounds.size() << "\n";
  std::ostream &out = sounds_cat.info();
  for (SoundInfo *sound : _sounds) {
    out << "Sound " << sound->name << "\n";
    indent(out, 2) << "index " << sound->index << "\n";
    indent(out, 2) << "chan " << sound->channel << "\n";
    indent(out, 2) << "soundlevel " << sound->soundlevel << "\n";
    indent(out, 2) << "vol min " << sound->volume_min << "\n";
    indent(out, 2) << "vol max " << sound->volume_max << "\n";
    indent(out, 2) << "pitch min " << sound->pitch_min << "\n";
    indent(out, 2) << "pitch max " << sound->pitch_max << "\n";
    indent(out, 2) << "dist mult " << sound->dist_mult << "\n";
    indent(out, 2) << "min dist " << sound->min_dist << "\n";
    indent(out, 2) << sound->waves.size() << "waves\n";
    for (const Wave &wave : sound->waves) {
      indent(out, 2) << "Wave " << wave.filename << "\n";
      if (wave.sound == nullptr) {
	indent(out, 4) << "Sound data NOT loaded\n";
      } else {
	indent(out, 4) << "Sound data is loaded\n";
      }
      indent(out, 4) << "loop start " << wave.loop_start << "\n";
      indent(out, 4) << "loop end " << wave.loop_end << "\n";
      indent(out, 4) << "spatialized " << wave.spatialized << "\n";
    }
  }
}

/**
 * Creates a sound instance for the given SoundInfo.  The returned sound is
 * an actual audio sound that can be played etc.
 */
PT(AudioSound) SoundManager::
create_sound(SoundInfo *info, bool spatial) {
  if (info == nullptr) {
    return null_sound;
  }

  Wave *wv = nullptr;
  if (info->waves.size() > 1u) {
    // Pick a random wave.
    int wave_index = _random.random_int((int)info->waves.size());
    wv = &info->waves[wave_index];
  } else if (!info->waves.empty()) {
    wv = &info->waves[0];
  }

  if (wv == nullptr) {
    // Got no wave?
    return null_sound;
  }

  PT(AudioSound) sound = get_wave_sound(wv, _sfx_manager);
  // Set properties on the sound that were specified on the sound info.
  sound->set_play_rate(info->pitch_min + _random.random_real_unit() * (info->pitch_max - info->pitch_min));
  sound->set_volume(info->volume_min + _random.random_real_unit() * (info->volume_max - info->volume_min));
  if (spatial) {
    // Set up spatialization properties.
    sound->set_3d_min_distance(info->min_dist);
    SteamAudioProperties sprops;
    sprops._enable_occlusion = true;
    sprops._bilinear_hrtf = false;
    sound->apply_steam_audio_properties(sprops);
  }

  return sound;
}

/**
 *
 */
PT(AudioSound) SoundManager::
create_sound_by_name(const std::string &name, bool spatial) {
  SoundInfo *info = get_sound_by_name(name);
  if (info == nullptr) {
    return null_sound;
  }
  return create_sound(info, spatial);
}

/**
 *
 */
PT(AudioSound) SoundManager::
get_wave_sound(Wave *wave, AudioManager *mgr) {
  if (wave->load_req != nullptr) {
    // We should already have an existing proxy if the load request
    // is active.
    if (sounds_cat.is_debug()) {
      sounds_cat.debug()
	<< "New proxy during async load for wave " << wave->filename << "\n";
    }
    assert(!wave->async_proxies.empty());
    PT(ProxyAudioSound) proxy = new ProxyAudioSound(wave->async_proxies[0].proxy);
    // Keep track of this sound proxy.
    wave->async_proxies.push_back({ proxy, mgr });
    return proxy;

  } else if (wave->sound != nullptr) {
    // We've got a sound data already, make a new instance.
    if (sounds_cat.is_debug()) {
      sounds_cat.debug()
	<< "Create wave instance " << wave->filename << "\n";
    }
    return mgr->get_sound(wave->sound);

  } else {
    // We haven't loaded the sound yet, load the sound asychronously
    // and in the meantime give them a proxy sound.  When the real
    // sound comes in we can apply any properties they set on the
    // proxy sound to the real thing.
    if (sounds_cat.is_debug()) {
      sounds_cat.debug()
	<< "Starting async load of sound data for wave " << wave->filename
	<< ", frame " << ClockObject::get_global_clock()->get_frame_count() << "\n";;
    }
    wave->load_req = new AudioLoadRequest(mgr, wave->filename, wave->spatialized);
    Loader *loader = Loader::get_global_ptr();
    loader->load_async(wave->load_req);
    PT(ProxyAudioSound) proxy = new ProxyAudioSound;
    proxy->set_loop_range(wave->loop_start, wave->loop_end);
    wave->async_proxies.push_back({ proxy, mgr });
    // Keep track of this wave so we can check if the load finished in update().
    _loading_waves.insert(wave);
    return proxy;
  }
}

/**
 *
 */
void SoundManager::
process_sound(KeyValues *sound_def) {
  PT(SoundInfo) pinfo = new SoundInfo;

  SoundInfo &info = *pinfo;
  info.name = sound_def->get_name();

  for (size_t i = 0; i < sound_def->get_num_keys(); ++i) {
    const std::string &key = sound_def->get_key(i);
    const std::string &value = sound_def->get_value(i);

    if (key == "channel") {
      info.channel = channel_from_script_name(value);

    } else if (key == "volume") {
      if (value == "VOL_NORM") {
	info.volume_max = info.volume_min = 1.0f;
      } else {
	// Volume format: min,max
	vector_string words;
	std::string trimmed = remove_spaces(value);
	tokenize(trimmed, words, ",");
	if (words.size() == 1u) {
	  // Single volume.
	  if (!string_to_float(words[0], info.volume_min)) {
	    sounds_cat.error()
	      << "Couldn't parse volume " << words[0] << " to float\n";
	  } else {
	    info.volume_max = info.volume_min;
	  }
	} else if (words.size() > 1u) {
	  // Min and max volume, chosen randomly.
	  if (!string_to_float(words[0], info.volume_min)) {
	    sounds_cat.error()
	      << "Couldn't parse min volume " << words[0] << " to float\n";
	  }
	  if (!string_to_float(words[1], info.volume_max)) {
	    sounds_cat.error()
	      << "Couldn't parse max volume " << words[1] << " to float\n";
	    info.volume_max = info.volume_min;
	  }
	}
      }

    } else if (key == "soundlevel") {
      info.soundlevel = soundlevel_from_script_name(value);
      info.dist_mult = soundlevel_to_dist_mult(info.soundlevel);
      if (info.dist_mult > 0.0f) {
	info.min_dist = 1.0f / info.dist_mult;
      } else {
	info.min_dist = 1000000.0f;
      }

    } else if (key == "wave") {
      info.waves.push_back(wave_from_script(value));

    } else if (key == "loopstart") {
      assert(!info.waves.empty());
      string_to_float(value, info.waves[0].loop_start);

    } else if (key == "loopend") {
      assert(!info.waves.empty());
      string_to_float(value, info.waves[0].loop_end);

    } else if (key == "pitch") {
      int pitch = pitch_from_script_name(value);
      if (pitch == PITCH_invalid) {
	std::string trimmed = remove_spaces(value);
	vector_string words;
	tokenize(trimmed, words, ",");
	if (words.size() > 1u) {
	  string_to_float(words[0], info.pitch_min);
	  string_to_float(words[1], info.pitch_max);
	} else {
	  string_to_float(words[0], info.pitch_min);
	  info.pitch_max = info.pitch_min;
	}
      } else {
	info.pitch_min = info.pitch_min = pitch;
      }

      info.pitch_min = pitch_to_play_rate(info.pitch_min);
      info.pitch_max = pitch_to_play_rate(info.pitch_max);
    }
  }

  // Handle random wave block.
  for (size_t i = 0; i < sound_def->get_num_children(); ++i) {
    KeyValues *child = sound_def->get_child(i);
    if (child->get_name() == "rndwave") {
      for (size_t j = 0; j < child->get_num_keys(); ++j) {
	std::string value = child->get_value(j);
	info.waves.push_back(wave_from_script(value));
      }
    }
  }

  info.index = (int)_sounds.size();
  _sounds.push_back(pinfo);
  _sounds_by_name[info.name] = pinfo;
}
